src/common/control.c

```diff
@@ -0,0 +1,231 @@
+            g_strcmp0(interface_name, SFMF_DBUS_INTERFACE) == 0 &&
+            sfmf_dbus_is_privileged(connection, sender)) {
+        if (g_strcmp0(method_name, "Abort") == 0) {
+            gboolean result = FALSE;
+            if (priv->callbacks && priv->callbacks->abort) {
+                result = priv->callbacks->abort(priv->callbacks_user_data);
+            }
+            g_dbus_method_invocation_return_value(invocation, g_variant_new("(b)", result));
+            return;
+        } else if (g_strcmp0(method_name, "GetProgress") == 0) {
+            g_dbus_method_invocation_return_value(invocation,
+                    g_variant_new("(sis)", g.progress.target ?: "", g.progress.progress,
+                        g.progress.phase ?: ""));
+            return;
+        }
+    }
+
+    g_dbus_method_invocation_return_dbus_error(invocation, SFMF_DBUS_INTERFACE ".MethodCallError",
+            "Invalid method call");
+}
+
+static const GDBusInterfaceVTable sfmf_control_vtable = {
+    sfmf_control_method_call_cb,
+    NULL,
+    NULL,
+};
+
+void sfmf_control_init(struct SFMF_Control_Callbacks *callbacks, void *user_data)
+{
+    g.callbacks = callbacks;
+    g.callbacks_user_data = user_data;
+
+    g.own_name = g_bus_own_name(G_BUS_TYPE_SYSTEM, SFMF_DBUS_NAME, G_BUS_NAME_OWNER_FLAGS_NONE,
+            sfmf_control_bus_acquired_cb, sfmf_control_name_acquired_cb, sfmf_control_name_lost_cb,
+            &g, NULL);
+
+    while (!g.name_acquired) {
+        sfmf_control_process();
+    }
+
+    GError *error = NULL;
+    GDBusNodeInfo *node_info = g_dbus_node_info_new_for_xml(""
+        "<!DOCTYPE node PUBLIC \"-//freedesktop//DTD D-BUS Object Introspection 1.0//EN\"\n"
+        "  \"http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd\">\n"
+        "<node name=\"" SFMF_DBUS_PATH "\">\n"
+        "  <interface name=\"" SFMF_DBUS_INTERFACE "\">\n"
+        "    <method name=\"Abort\">\n"
+        "      <arg name=\"result\" type=\"b\" direction=\"out\" />\n"
+        "    </method>\n"
+        "    <method name=\"GetProgress\">\n"
+        "      <arg name=\"subvolume\" type=\"s\" direction=\"out\" />\n"
+        "      <arg name=\"progress\" type=\"i\" direction=\"out\" />\n"
+        "      <arg name=\"phase\" type=\"s\" direction=\"out\" />\n"
+        "    </method>\n"
+        "    <signal name=\"Progress\">\n"
+        "      <arg name=\"subvolume\" type=\"s\" />\n"
+        "      <arg name=\"progress\" type=\"i\" />\n"
+        "      <arg name=\"phase\" type=\"s\" />\n"
+        "    </signal>\n"
+        "  </interface>\n"
+        "</node>\n"
+    "", &error);
+    if (!node_info) {
+        SFMF_FAIL("Could not compile D-Bus XML: %s\n", error->message);
+        g_error_free(error);
+    }
+
+    g.object_registration = g_dbus_connection_register_object(g.connection, "/",
+            g_dbus_node_info_lookup_interface(node_info, SFMF_DBUS_INTERFACE),
+            &sfmf_control_vtable, &g, NULL, &error);
```

Comment:
`SFMF_FAIL()` will exit there, but I'll put that assignment into the `} else {` of the preceding `if (!node_info)` just to be sure.